packages/@expo/cli/src/start/server/metro/createExpoStickyResolver.ts

```diff
@@ -0,0 +1,172 @@
+import type { SearchOptions as AutolinkingSearchOptions } from 'expo-modules-autolinking/exports';
+import type { ResolutionContext } from 'metro-resolver';
+
+import type { StrictResolverFactory } from './withMetroMultiPlatform';
+import type { ExpoCustomMetroResolver } from './withMetroResolvers';
+
+const debug = require('debug')('expo:start:server:metro:sticky-resolver') as typeof console.log;
+
+const AUTOLINKING_PLATFORMS = ['android', 'ios', 'web'] as const;
+type AutolinkingPlatform = (typeof AUTOLINKING_PLATFORMS)[number];
+
+/** Converts a list of module names to a regex that may either match bare module names or sub-modules of modules */
+const dependenciesToRegex = (dependencies: string[]) =>
+  new RegExp(`^(${dependencies.join('|')})($|/.*)`);
+
+/** Creates a function to load a dependency of the `expo` package */
+const createExpoDependencyLoader = <T>(request: string) => {
+  let mod: T | undefined;
+  return (): T => {
+    if (!mod) {
+      const expoPath = require.resolve('expo/package.json');
+      const autolinkingPath = require.resolve(request, {
+        paths: [expoPath],
+      });
+      return (mod = require(autolinkingPath));
+    }
+    return mod;
+  };
+};
+
+const getAutolinkingModule = createExpoDependencyLoader<
+  typeof import('expo-modules-autolinking/exports')
+>('expo-modules-autolinking/exports');
```

Comment:
we can export `expo/autolinking` from `expo-modules-autolinking/exports`, so that we don't necessarily to resolve `expo -> expo-modules-autolinking/exports`